Fix godeps

soltysh opened the pull request:
This PR does 3 things:
1. fix godeps, further down details about each fixed godep (result of running `hack/godep-save.sh`)
2. adds preloading `google.golang.org/cloud` to `godep-restore.sh`, for some reason it fails regular restore process 
3. adds `hack/godep-restore.sh` invocation to `make verify` so that godeps are check on every PR

Godeps details:
- `cloud.google.com/go/storage` migrated to 
- remove `github.com/aws/aws-sdk-go/private/protocol` it's just a dir currently
- add `github.com/codegangsta/negroni` used by k8spetstore example
- add missing `github.com/docker/distribution/testutil`
- add missing `github.com/google/cadvisor/info/v1/test`
- add missing `github.com/jteeuwen/go-bindata/go-bindata`, used by k8s
- add missing `github.com/ncw/swift/swifttest`, used by docker/distribution
- add missing `github.com/onsi/ginkgo/ginkgo` and subdirs
- add missing `github.com/openshift/source-to-image/pkg/util/status`
- add missing `github.com/xyproto/simpleredis` used by k8spetstore example
- add missing `google.golang.org/appengine` and subdirs used by k8s
- add missing `google.golang.org/cloud/storage` ad subdirs used by docker/distribution
- add missing `google.golang.org/grpc/credentials/oauth` used by etcd
- add missing `k8s.io/client-go` and subdirs
- add missing `k8s.io/kubernetes` subdirs

@ncdc fyi
@smarterclayton || @liggitt ptal

soltysh commented:
[test]

ncdc commented on Makefile:
```diff
@@ -52,6 +52,7 @@ build-tests:
 #   make check
 check: | build verify
 	$(MAKE) test-unit test-cmd -o build -o verify
+	hack/godep-restore.sh
```
We should only run this if Godeps or vendor is in the patch

soltysh replied:
That's reasonable. I'll update this appropriately.

ncdc commented:
@soltysh what is this k8s.io/kubernetes bump for?

ncdc commented:
If you need to bump multiple godeps, can you do 1 per commit?

soltysh commented:
> @soltysh what is this k8s.io/kubernetes bump for?
> If you need to bump multiple godeps, can you do 1 per commit?

The bump is everything named in the description, I can split that into multiple, but most is k8s dependent. So maybe doing per specific origin dep, so that we don't have too many commits. Besides a lot of those updates just `Godeps.json`, without even touching `vendor/` dir.

ncdc commented:
From a rebase perspective, it's much easier for me if each godep is bumped in its own commit. I had trouble getting the openshift-1.4 branch working because there was a PR where 1 commit bumped multiple godeps and Godeps.json actually wasn't correct (I realize running a verification test during the PR would have caught that).

stevekuznetsov commented:
> From a rebase perspective, it's much easier for me if each godep is bumped in its own commit.

Seems like a reasonable burden -- let's enforce it in the `commitchecker` linter

liggitt commented:
>> From a rebase perspective, it's much easier for me if each godep is bumped in its own commit.
>
> Seems like a reasonable burden -- let's enforce it in the commitchecker linter

Not sure... typically a rebase brings in lots of bumps to things we don't reference directly... one commit per transitive dependency is noisy

ncdc commented:
Really what I'm looking for is a way to ensure that for non UPSTREAM PRs, where we're actually doing a bump, that the shas in Godeps.json are all updated appropriately. It took me a while to find the github.com/opencontainers/runc commit we were actually using because its sha wasn't updated in a cadvisor bump commit like it should have been.

ncdc commented:
I had to guess at which runc commit we were using, then run `diff -rqwb vendor/github.com/opencontainers/runc $GOPATH/src/github.com/opencontainers/runc`, look for deltas, and keep moving the runc HEAD in my $GOPATH until I found the right sha.

stevekuznetsov commented:
> one commit per transitive dependency is noisy

If we're not generating them by hand, I'm fairly certain nobody is _reading_ them by hand, so I don't really see the negative of noise ... see also: @openshift-bot making a commit bump for every web-console merge.

liggitt commented:
> If we're not generating them by hand, I'm fairly certain nobody is reading them by hand

you assume wrong

stevekuznetsov commented:
> you assume wrong

On which end? Generating them, that's one short script away from possible. Reading them... I perfer novels to our git history, but to each their own :)

soltysh commented:
From my previous experience I agree with @liggitt one common bump for transitive deps under the umbrella of direct origin dep. This is reasonable approach and easy to understand/explain. I'll update this PR with that approach in mind. Web-console is a direct dep, that's why we'll always get direct bump message for that.

soltysh commented:
> Seems like a reasonable burden -- let's enforce it in the commitchecker linter

Doesn't make sense, since `commitchecker` ignores all bumps, and this is what we actually care most of the times. Although running this with all `UPSTREAM *` commits is desirable as well, just in case somebody will mess up. Generally everytime `vendor/` dir is touched this should be run, imho. And that's the approach I'm working on right now.

soltysh commented:
I'll go with extending `commitchecker` it looks like writing such a script would be over engineering, esp. that we have all the necessary checks in place already.

soltysh commented:
I've split the bump into multiple direct origin's dependencies. Only the k8s one is a big bigger, the reaming are usually one dep per commit, or even just updating `Godeps.json`. I hope that looks better to you @ncdc and it's ok with you @liggitt ?
@stevekuznetsov I've added new validator for the commitchecker, but since commitchecker is being called in travis I had to enable this through env variable and add 2nd invocation in `make check` that forces checking godeps. 
Let's see what jenkins do with that now.

stevekuznetsov commented on tools/rebasehelpers/util/git.go:
```diff
@@ -98,12 +98,15 @@ func (c Commit) GodepsReposChanged() ([]string, error) {
 type File string
 
-func (f File) HasGodepsChanges() bool {
+func (f File) HasGodepsChanges(includingGodeps bool) bool {
```
Passing in a `bool` just to switch on it inside of the method is bad. You have two different methods, with two different actions, let's not overload `HasGodepsChanges` to try to handle both. Also, the new method signature is very hard to read:

```go
if !file.HasGodepsChanges(false) {
```

stevekuznetsov replied:
```go
func (f File) ChangesVendoredCode() bool {
    return strings.HasPrefix(string(f), "Godeps/_workspace") || strings.HasPrefix(string(f), "vendor")
}

func (f File) ChangesVendorManifest() bool {
    return f == "Godeps/Godeps.json"
}
```

soltysh replied:
sgtm, will update.

soltysh replied:
I went with `HasVendoredCodeChanges` and `HasGodepsChanges` since we're already using godeps elsewhere for Godeps.json so I kept that in sync. I guess that's reasonable approach.

liggitt commented on tools/rebasehelpers/commitchecker/commitchecker.go:
```diff
@@ -36,8 +36,13 @@ func main() {
+			errs = append(errs, err.Error())
+		}
+	}
+	if len(os.Getenv("GODEP_RESTORE")) > 0 {
```
We've definitely run into cases when refactoring that would have been extremely painful to make satisfy a strict bump verifier. 

 any time you are querying history for bumps and carry patches, you should really be doing it with git history of the actual sha referenced in the godep content, not based on commit titles… those are a convenience, not authoritative

soltysh replied:
The only reason I'm not enabling this always is because commitchecker is run inside travis and I doubt it'll succeed there. The env is always on in `make check`, which is always run in jenkins. Then it looks at either modified `Godeps/_workspace`, `vendor/` dirs or `Godeps/Godeps.json`, if any of these is touched it'll run `hack/godep-restore.sh`.

liggitt replied:
No, I mean the check for changes to a vendored package since a bump should be done using git history, not commit titles

soltysh commented:
Comments addressed and I rebased the PR. @stevekuznetsov @liggitt ptal

stevekuznetsov commented on tools/rebasehelpers/util/git.go:
```diff
@@ -59,6 +59,18 @@ func (c Commit) DeclaredUpstreamRepo() (string, error) {
+	return false
+}
+
+// HasGodepsChanges verifies if the commit has any changes to Godeps/Godeps.json file.
```
Nit, but `GodepsChanges` is overloaded, especially as before it meant changes to the vendored files, which is why I called out `VendorManifest` explicitly. Not a blocker.

soltysh replied:
As I said, it's also being used elsewhere wrt to `Godeps/Godeps.json` - `GodepsReposChanged`, specifically. That's why I went with this name and most importantly added godoc, so it's clear what it does 😄

stevekuznetsov replied:
I see your comments now. Blame GitHub for not giving me the notification :)

soltysh commented:
@stevekuznetsov @liggitt @ncdc if no other objections I'd like to get this one merged today.

liggitt commented:
I would *not* expect running `make check` to call `godep restore` and touch stuff outside the origin repo

ncdc commented:
+1, `godep restore` should be something you do intentionally and something you're aware of, instead of hidden behind `make check`.

soltysh commented:
How else can we verify `godep restore` is working properly? Should we have a separate check that will be explicitly called in our tests?

ncdc commented:
Yes, as part of the jenkins run

soltysh commented:
That makes sense, I'll update this first and the once it merges vagrant-openshift.

ncdc commented:
I'm not sure it needs a vagrant-openshift change? We alternatively could modify the jenkins build script?

soltysh commented:
jenkins build script calls vagrant actions, it needs to land in the action related to origin testing, [here](https://github.com/openshift/vagrant-openshift/blob/5d517fc9dc8a78b05a5ab80a131de180b02dd96d/lib/vagrant-openshift/action/run_origin_tests.rb#L93) will be the right place, imho.

soltysh commented:
Updated. I've created `make verify-godeps` which calls that new mode.

stevekuznetsov commented:
Would rather see `make check` do it if a specific envar was present, than having to change `vagrant-openshift`.

soltysh commented:
The godep verification will be enabled through `VERIFY_GODEPS` environment variable that needs to be explicitly passed to jenkins. We'll reconfigure our job so that it sets this variable in origin tests, specifically when invoking `make check -> hack/verify-upstream-commits.sh`. Without it explicitly set, no checks are done. I hope that satisfies you both @liggitt @ncdc. With that in mind any other objections to tag this for merge?

stevekuznetsov commented:
> With that in mind any other objections to tag this for merge?

If it's green-lit, update the job and get a green test on this before merging.

liggitt commented:
no objection in principle. slight preference for a scarier envvar that makes it clear that godep restore is being called, like `RESTORE_AND_VERIFY_GODEPS`

liggitt commented:
do we lock to a particular version of godep? should we, if we're going to block merges on godep issues?

soltysh commented:
Yes `hack/godep-restore.sh` locks to godep v63, see [here](https://github.com/openshift/origin/blob/0836881720778991bb911c4158621070acaf0500/hack/godep-restore.sh#L65). I'll change the name.

soltysh commented:
I've updated jenkins check job, let's give it a go with the new variable.

[test]

soltysh commented:
@stevekuznetsov the current approach is not good :/ apparently the way we invoke `make check` is not inside a git repository (the rsync looks bogus, I guess), which basically means `hack/verify-upstream-commits.sh` is skipped!
```
hack/verify-upstream-commits.sh
SKIPPED: Not a Git repository
```
That means only travis will fail if somebody will be interested in committing wrong upstream commit, and we both know that travis isn't what folks look after 😉 . Let's talk about possible solution when you're online.

soltysh commented:
This is waiting for #12056 to be merged first, then I'll update jenkins job and re-run tests. Only if that succeeds I'll merge this.

soltysh commented:
#12056 merged, I've updated jenkins job and rebased this PR. Waiting for tests results.

soltysh commented:
Tests are green so I'm merging this as is, but I'll keep an eye on any failures in there. This is needed for Andy for the rebase, asap.

[merge]

liggitt commented:
the jenkins change means the job can't run on non-master branches - https://github.com/openshift/origin/pull/12070#issuecomment-264486727

soltysh commented:
@liggitt fixed in https://github.com/openshift/origin/pull/12117
